WordPress/gutenberg
# Add useQuerySelect to core/data

## Description

This PR proposes a new hook called `useResolveSelect`. It is used like this:

```js
const { data, isResolving, hasFinished } = useResolveSelect(
	( resolve ) => resolve( coreStore ).getEntityRecord( 'postType', 'page', pageId ),
	[ pageId ]
};
```

The rationale is to make snippets like below more readable and faster to write – I find myself looking up the correct way to work with resolution status every time:
```js
const { data, isResolving, hasFinished } = useSelect(
	( select ) => {
		const {
			getEntityRecord,
			isResolving,
			hasFinishedResolution
		} = select( coreStore );
		const args = [ 'postType', 'page', pageId ];
		return {
			data: getEntityRecord( ...args ),
			isResolving: isResolving( 'getEntityRecord', args ),
			hasFinished: hasFinishedResolution( 'getEntityRecord', args ),
		}
	},
	[ pageId ]
};
```

In https://github.com/WordPress/gutenberg/pull/38135 I also drafted a higher-level hook called `useEntityRecord` building on top of this PR:

```js
const { data, isResolving, hasFinished } = useEntityRecord( 'postType', 'page', pageId );
```

@youknowriad's work on the suspense API support in https://github.com/WordPress/gutenberg/pull/37261 is related, but I see it as a separate API that doesn't diminish the value of having `useResolveSelect`.

## Testing steps

This is a new hook so there's not much to test in the UI. Here's an alternative test plan:

1. Confirm the code makes sense
2. Confirm the unit tests pass and make sense
3. Read the included documentation and examples and confirm they make sense

cc @kevin940726 @talldan @gziolo @draganescu @ellatrix @noisysocks @jsnajdr @getdave
